java/anaplan-connect-library/ac-auth/ac-auth-default/src/main/java/com/anaplan/client/auth/DeviceAuthenticator.java

```diff
@@ -0,0 +1,211 @@
+import com.anaplan.client.CryptoUtil;
+import com.anaplan.client.Utils;
+import com.anaplan.client.api.AnaplanAuthenticationAPI;
+import com.anaplan.client.dto.DeviceCodeInfo;
+import com.anaplan.client.dto.OauthTokenInfo;
+import com.anaplan.client.exceptions.AnaplanAPIException;
+import com.anaplan.client.transport.ConnectionProperties;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.security.Key;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.CertificateException;
+import java.util.Base64;
+import java.util.Objects;
+
+/**
+ * Authenticate by device
+ */
+public class DeviceAuthenticator extends AbstractAuthenticator {
+
+  public static final String ROTATABLE = "rotatable";
+
+  private static final Logger LOG = LoggerFactory.getLogger(DeviceAuthenticator.class);
+  private static final String GRANT_TYPE = "urn:ietf:params:oauth:grant-type:device_code";
+  private static final String DEVICE_SCOPE = "openid profile email offline_access";
+  private static final String REFRESH_TOKEN = "refresh_token";
+  private static final String JKS = ".jks";
+  private final String refreshTokenKeyStorePath;
+  private final String refreshTokenKeyStoreName;
+  private final char[] keystorePass;
+  private OauthTokenInfo oauthTokenInfo;
+
+  public DeviceAuthenticator(ConnectionProperties connectionProperties,
+                             AnaplanAuthenticationAPI authClient) {
+    super(connectionProperties, authClient);
+    refreshTokenKeyStoreName = String.format("ks_%s", Utils.bytesToHex(Utils.createHash(connectionProperties.getClientId())));
+    String keystoreDir = System.getProperty("user.home");
+    if (StringUtils.isNotBlank(System.getenv("AC_OAUTH_KEYSTORE_DIR"))) {
+      keystoreDir = System.getenv("AC_OAUTH_KEYSTORE_DIR");
+    }
+    refreshTokenKeyStorePath = keystoreDir + FileSystems.getDefault().getSeparator() + refreshTokenKeyStoreName + JKS;
+    keystorePass = CryptoUtil.encrypt(this.connectionProperties.getClientId()).toCharArray();
+  }
+
+  @Override
+  public byte[] authenticate() {
+    LOG.info("Authenticating via Device...");
+    String refreshToken = getDecodedRefreshToken();
+    if (connectionProperties.isForceRegister()) {
+      clearRefreshTokenEntry();
+      deviceRegistration();
+      return authToken;
+    } else if (StringUtils.isEmpty(refreshToken)) {
+      deviceRegistration();
+      return authToken;
+    } else {
+      synchronized (this) {
+        return refreshToken();
+      }
+    }
+  }
+
+  public void clearRefreshTokenEntry() {
+    LOG.info("Deleting already existing JKS and re-registering the device...");
```

Comment:
Can you add Java Dc for this in ACL and here as well?

Reply:
Java DC has been added